data_diff/dbt.py

```diff
@@ -0,0 +1,313 @@
+
+        for model in models:
+            diff_vars = self._get_diff_vars(dbt_parser, config_prod_database, config_prod_schema, model, datasource_id)
+
+            if is_cloud and len(diff_vars.primary_keys) > 0:
+                self._cloud_diff(diff_vars)
+            elif is_cloud:
+                rich.print(
+                    f"[red]"
+                    + ".".join(diff_vars.dev_path)
+                    + " <> "
+                    + ".".join(diff_vars.prod_path)
+                    + "[/] \n"
+                    + "Skipped due to missing primary-key tag\n"
+                )
+
+            if not is_cloud and len(diff_vars.primary_keys) == 1:
+                self._local_diff(diff_vars)
+            elif not is_cloud:
+                rich.print(
+                    f"[red]"
+                    + ".".join(diff_vars.dev_path)
+                    + " <> "
+                    + ".".join(diff_vars.prod_path)
+                    + "[/] \n"
+                    + "Skipped due to missing primary-key tag or multi-column primary-key (unsupported for non --cloud diffs)\n"
+                )
+
+        rich.print("Diffs Complete!")
+
+    @staticmethod
+    def _get_diff_vars(dbt_parser, config_prod_database, config_prod_schema, model, datasource_id) -> DiffVars:
+        dev_database = model.database
+        dev_schema = model.schema_
+        primary_keys = dbt_parser.get_primary_keys(model)
+
+        if config_prod_database is None:
+            prod_database = dev_database
+        else:
+            prod_database = config_prod_database
+
+        if config_prod_schema is None:
+            prod_schema = dev_schema
+        else:
+            prod_schema = config_prod_schema
+
+        if dbt_parser.requires_upper:
+            dev_qualified_list = [x.upper() for x in [dev_database, dev_schema, model.name]]
+            prod_qualified_list = [x.upper() for x in [prod_database, prod_schema, model.name]]
+            primary_keys = [x.upper() for x in primary_keys]
+        else:
+            dev_qualified_list = [dev_database, dev_schema, model.name]
+            prod_qualified_list = [prod_database, prod_schema, model.name]
+        return DiffVars(dev_qualified_list, prod_qualified_list, primary_keys, datasource_id, dbt_parser.connection)
+
+    @staticmethod
+    def _local_diff(diff_vars: DiffVars):
+        column_diffs_str = ""
+        dev_qualified_string = ".".join(diff_vars.dev_path)
+        prod_qualified_string = ".".join(diff_vars.prod_path)
+        primary_key = diff_vars.primary_keys[0]
+
+        table1 = connect_to_table(diff_vars.connection, dev_qualified_string, primary_key)
+        table2 = connect_to_table(diff_vars.connection, prod_qualified_string, primary_key)
+
+        table1_columns = list(table1.get_schema())
+        try:
+            table2_columns = list(table2.get_schema())
+        # TODO add a unit test for scenario
+        except Exception as e:
```

Comment:
Is there a specialized exception class escalating from the database connectors? Because `Exception` can be too generic — e.g. a division by zero or type mismatches. That will be highly confusing to see an error about the table access in such cases.

Reply:
Unfortunately that doesn't seem to be the case. There is [this](https://github.com/datafold/sqeleton/blob/ffc4afe80cdd54b9ee29d2aa429346af7cce5a25/sqeleton/databases/base.py#L391) exception, but exceptions can be raised by the various underlying connection packages before that. And they aren't being routed to an exception type that spans them

For now, maybe I can log the exception every time so the user can interpret